### PR TITLE
Improved `get_type_hints()` for classes with forward references

### LINKED ISSUE #1004: from_type() does not resolve forward references outside of function annotations
The sample code:
```python
import typing as T

import hypothesis as H
import hypothesis.strategies as S

class Tree(T.NamedTuple):
    val: int
    l: T.Optional["Tree"]
    r: T.Optional["Tree"]
    def size(self):
        return 1 + (self.l.size() if self.l else 0) + (self.r.size() if self.r else 0)

@H.given(t = S.infer)
def test_tree(t: Tree):
    if t.size() > 3:
        assert False
        H.note(t)
    else:
        assert True
```

(Part of) the stack trace:
```
platform darwin -- Python 3.6.1, pytest-3.2.1, py-1.4.34, pluggy-0.4.0
rootdir: /Users/desmond/Documents/code/code-python/Try1/tests, inifile:
plugins: xonsh-0.5.12, hypothesis-3.38.9
collected 4 items
foo_test.py ...F
foo_test.py:36 (test_tree)
args = (<class 'foo_test.Tree'>,), kwargs = {}, kwargs_cache_key = set()
cache_key = (<function from_type at 0x112216a60>, ((<class 'type'>, <class 'foo_test.Tree'>),), frozenset())
result = builds(Tree, l=builds(lazy_error), r=builds(lazy_error), val=integers())

    @proxies(fn)
    def cached_strategy(*args, **kwargs):
        kwargs_cache_key = set()
        try:
            for k, v in kwargs.items():
                kwargs_cache_key.add((k, convert_value(v)))
        except TypeError:
            return fn(*args, **kwargs)
        cache_key = (
            fn,
            tuple(map(convert_value, args)), frozenset(kwargs_cache_key))
        try:
>           return STRATEGY_CACHE[cache_key]

/Users/desmond/anaconda3/lib/python3.6/site-packages/hypothesis/strategies.py:107: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = {(<function booleans at 0x112763e18>, (), frozenset()): booleans(), (<function fractions at 0x112216d90>, ((<class 'No...(tuples(), fixed_dictionaries({})).map(lambda value: target(*value[0], **value[1])),
 'val': WideRangeIntStrategy()}))}
key = (<function from_type at 0x112216a60>, ((<class 'type'>, <class 'foo_test.Tree'>),), frozenset())

    def __getitem__(self, key):
>       i = self.keys_to_indices[key]
E       KeyError: (<function from_type at 0x112216a60>, ((<class 'type'>, <class 'foo_test.Tree'>),), frozenset())

```

### PR DESCRIPTION
**Pull request by Zac-HD**
This was prompted by some work on https://github.com/Zac-HD/hypothesmith/issues/2.  Happily, it also fixes #1004 - since I last prodded at that issue the typing ecosystem has improved and we're *mostly* 3.6+ and so a fairly simple approach now works :tada:

### CODE REVIEW

File: hypothesis-python/src/hypothesis/internal/compat.py [hunk 1]
```diff
@@ -125,14 +125,24 @@ def get_type_hints(thing):
 else:
 
     def get_type_hints(thing):
-        if inspect.isclass(thing) and not hasattr(thing, "__signature__"):
-            if is_typed_named_tuple(thing):
-                # Special handling for typing.NamedTuple
-                return thing._field_types  # type: ignore
-            thing = thing.__init__  # type: ignore
+        """Like the typing version, but tries harder and never errors.
```
Comment (rsokl) [inline hypothesis-python/src/hypothesis/internal/compat.py:128]: Now _this_ is a docstring. Can we have these sorts of docs from all of `hypothesis.internal` please?

File: hypothesis-python/src/hypothesis/internal/compat.py [hunk 2]
```diff
@@ -125,14 +125,24 @@ def get_type_hints(thing):
...
+
+        Tries harder: if the thing to inspect is a class and raises an error or we
+        find no hints, this function will then inspect the __init__ method.  This
+        often helps with user-defined classes on older versions of Python.
```
Comment (rsokl) [inline hypothesis-python/src/hypothesis/internal/compat.py:132]: sorry for the nit, but this actually was hard to parse at first

```suggestion
        Tries harder: if the thing to inspect is a class, but it raises an error or we
        find no hints, then this function will then inspect the __init__ method.  This
        often helps with user-defined classes on older versions of Python.
```

File: hypothesis-python/src/hypothesis/internal/compat.py (hunk 1 above)
Comment (Zac-HD) [reply hypothesis-python/src/hypothesis/internal/compat.py:128]: :joy: Be the change you want to see in the world!
